include/crocoddyl/multibody/costs/state.hxx

```diff
@@ -150,9 +150,23 @@ void CostModelStateTpl<Scalar>::calcDiff(const boost::shared_ptr<CostDataAbstrac
   CostDataStateTpl<Scalar>* d = static_cast<CostDataStateTpl<Scalar>*>(data.get());
   state_->Jdiff(xref_, x, data->Rx, data->Rx, second);
   activation_->calcDiff(data->activation, data->r);
-  data->Lx.noalias() = data->Rx.transpose() * data->activation->Ar;
-  d->Arr_Rx.noalias() = data->activation->Arr * data->Rx;
-  data->Lxx.noalias() = data->Rx.transpose() * d->Arr_Rx;
+
+  const boost::shared_ptr<pinocchio::ModelTpl<Scalar> > pin_model = state_->get_pinocchio();
+  typedef Eigen::Block<MatrixXs> MatrixBlock;
+  for (pinocchio::JointIndex i = 1; i < (pinocchio::JointIndex)pin_model->njoints; ++i) {
+    const MatrixBlock& RxBlock =
+        data->Rx.block(pin_model->idx_vs[i], pin_model->idx_vs[i], pin_model->nvs[i], pin_model->nvs[i]);
+
+    data->Lx.segment(pin_model->idx_vs[i], pin_model->nvs[i]).noalias() =
+        RxBlock.transpose() * data->activation->Ar.segment(pin_model->idx_vs[i], pin_model->nvs[i]);
+
+    data->Lxx.block(pin_model->idx_vs[i], pin_model->idx_vs[i], pin_model->nvs[i], pin_model->nvs[i]).noalias() =
+        RxBlock.transpose() *
```

Comment:
This implementation is much more efficient than the one we had before, since it uses the sparsity of the Rx and Arr matrices. For all joints except the freeflyer, the dimension of Rxblock would be (1,1) and dimension of Arr would be (1,1). So for the instance I'm going to say that it is okay. But this is not perfect.

A better solution does exist, and eventually this implementation would need to go inside pinocchio, whereit would be the continuation of https://github.com/stack-of-tasks/pinocchio/pull/1169, but with the `APPLY_ON_THE_LEFT` operators, and `dDifference`algorithm.

Reply:
Are you planning to integrate the Pinocchio solution in this PR?

I am OK if you prefer to integrate it in a next iteration.

Reply:
It would have to be later, unfortunately. I can't work much more on this PR.

Reply:
I'll close this.